src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,287 @@
+                }
+            }
+
+            if (!done) {
+                // Now traverse the second branch of the utxo we have arrived at.
+                selection.at(depth).second = true;
+
+                // These were always included first, try excluding now
+                selection.at(depth).first = false;
+                value_ret -= utxo_pool.at(depth).txout.nValue;
+                curr_waste -= input_fee_vec[depth] - long_term_fee_vec[depth];
+                ++depth;
+            }
+        }
+        --remaining_tries;
+    }
+
+    // Check for solution
+    if (best_selection.empty()) {
+        return false;
+    }
+
+    // Set output set
+    value_ret = 0;
+    for (unsigned int i = 0; i < best_selection.size(); ++i) {
+        if (best_selection.at(i).first) {
+            out_set.insert(utxo_pool.at(i));
+            fee_ret += input_fee_vec.at(i);
+            value_ret += utxo_pool.at(i).txout.nValue;
+        }
+    }
+
+    return true;
+}
+
+static void ApproximateBestSubset(const std::vector<CInputCoin>& vValue, const CAmount& nTotalLower, const CAmount& nTargetValue,
+                                  std::vector<char>& vfBest, CAmount& nBest, int iterations = 1000)
+{
+    std::vector<char> vfIncluded;
+
+    vfBest.assign(vValue.size(), true);
+    nBest = nTotalLower;
+
+    FastRandomContext insecure_rand;
+
+    for (int nRep = 0; nRep < iterations && nBest != nTargetValue; nRep++)
+    {
+        vfIncluded.assign(vValue.size(), false);
+        CAmount nTotal = 0;
+        bool fReachedTarget = false;
+        for (int nPass = 0; nPass < 2 && !fReachedTarget; nPass++)
+        {
+            for (unsigned int i = 0; i < vValue.size(); i++)
+            {
+                //The solver here uses a randomized algorithm,
+                //the randomness serves no real security purpose but is just
+                //needed to prevent degenerate behavior and it is important
+                //that the rng is fast. We do not use a constant random sequence,
+                //because there may be some privacy improvement by making
+                //the selection random.
+                if (nPass == 0 ? insecure_rand.randbool() : !vfIncluded[i])
+                {
+                    nTotal += vValue[i].txout.nValue;
+                    vfIncluded[i] = true;
+                    if (nTotal >= nTargetValue)
+                    {
+                        fReachedTarget = true;
+                        if (nTotal < nBest)
+                        {
+                            nBest = nTotal;
+                            vfBest = vfIncluded;
+                        }
+                        nTotal -= vValue[i].txout.nValue;
+                        vfIncluded[i] = false;
+                    }
+                }
+            }
+        }
+    }
+}
+
+bool KnapsackSolver(std::vector<CInputCoin>& utxo_pool, const CAmount& nTargetValue, std::set<CInputCoin>& out_set, CAmount& value_ret)
```

Comment:
Note for reviewers: Again this code is mostly not new. It's moved from the previous `SelectCoinsMinConf` implementation in wallet.cpp